src/i_winmusic.c

```diff
@@ -1380,28 +1326,61 @@ static DWORD WINAPI PlayerProc(void)
 {
     while (true)
     {
-        if (WaitForSingleObject(hBufferReturnEvent, INFINITE) == WAIT_OBJECT_0)
+        if (WaitForSingleObject(hBufferReturnEvent, INFINITE) != WAIT_OBJECT_0)
         {
-            EnterCriticalSection(&CriticalSection);
+            continue;
+        }
 
-            switch (win_midi_state)
-            {
-                case STATE_STOPPED:
-                    SetEvent(hStoppedEvent);
-                    break;
+        EnterCriticalSection(&CriticalSection);
 
-                case STATE_EXIT:
-                    LeaveCriticalSection(&CriticalSection);
-                    return 0;
+        buffer.position = 0;
 
-                default:
+        switch (win_midi_state)
+        {
+            case STATE_STARTUP:
+                ResetDevice();
+                StreamOut();
+                song.rpg_loop = IsRPGLoop();
+                win_midi_state = STATE_PLAYING;
+                break;
+
+            case STATE_SHUTDOWN:
+                // Send notes/sound off prior to reset to prevent volume spikes.
+                SendNotesSoundOff();
+                ResetDevice();
+                StreamOut();
+                win_midi_state = STATE_EXIT;
+                break;
+
+            case STATE_EXIT:
+                LeaveCriticalSection(&CriticalSection);
+                return 0;
+
+            case STATE_PLAYING:
+                if (update_volume)
+                {
+                    UpdateVolume();
+                    StreamOut();
+                    update_volume = false;
                     break;
-            }
+                }
+                FillBuffer();
+                break;
 
-            FillBuffer();
+            case STATE_STOPPING:
+                // Send notes/sound off to prevent hanging notes.
+                SendNotesSoundOff();
+                StreamOut();
+                WaitForSingleObject(hBufferReturnEvent, INFINITE);
+                win_midi_state = STATE_STOPPED;
+                SetEvent(hStoppedEvent);
```

Comment:
How about this change?
```diff
diff --git a/src/i_winmusic.c b/src/i_winmusic.c
index cac7a21c..743504ec 100644
--- a/src/i_winmusic.c
+++ b/src/i_winmusic.c
@@ -1371,8 +1371,10 @@ static DWORD WINAPI PlayerProc(void)
                 // Send notes/sound off to prevent hanging notes.
                 SendNotesSoundOff();
                 StreamOut();
-                WaitForSingleObject(hBufferReturnEvent, INFINITE);
                 win_midi_state = STATE_STOPPED;
+                break;
+
+            case STATE_STOPPED:
                 SetEvent(hStoppedEvent);
                 break;
 
```

Reply:
If those lines are added (reverting the change in this PR), it's possible for `SetEvent(hStoppedEvent)` to occur AFTER `ResetEvent(hStoppedEvent)` is called in the `StreamStop()` function. That means the `WaitForSingleObject()` inside `I_WIN_StopSong()` would be immediately signaled, canceling `SendNotesSoundOff()` which can lead to hanging notes.

Reply:
What if we move `ResetEvent(hStoppedEvent)` to `I_WIN_StopSong()`?
```diff
diff --git a/src/i_winmusic.c b/src/i_winmusic.c
index cac7a21c..445fa91f 100644
--- a/src/i_winmusic.c
+++ b/src/i_winmusic.c
@@ -1425,7 +1425,6 @@ static void StreamStop(void)
     }
 
     ResetEvent(hBufferReturnEvent);
-    ResetEvent(hStoppedEvent);
 }
 
 static void GetDevices(void)
@@ -1561,6 +1560,7 @@ static void I_WIN_StopSong(void *handle)
 
     LeaveCriticalSection(&CriticalSection);
 
+    ResetEvent(hStoppedEvent);
     StreamStart();
     WaitForSingleObject(hStoppedEvent, PLAYER_THREAD_WAIT_TIME);
     StreamStop();
```

Reply:
It has the same problem, but now it's moved down a few lines.

Reply:
Oh, I think I finally get it. It got confusing 😄 Thanks for the explanation.

Reply:
I went back to your suggestion after doing a complete review of the interaction between the main thread and midi thread. With the extra changes, I think this is more robust now.

Reply:
Yes, I like the decision to call `StreamStart()`/`StreamStop()` functions from the critical section, it feels more correct. Thanks!